File: src/typescript/message-bus.ts
export interface ParsedEvent {
  type: string;
  data: Record<string, any>;
  stream?: AsyncIterable<string>;
  raw?: string;
}

export type EventHandler = (event: ParsedEvent, bus: MessageBus) => Promise<void>;

export class MessageBus {
  private _handlers: Map<string, EventHandler[]> = new Map();
  private _maxDepth: number;
  private _depth = 0;

  constructor(maxDepth = 10) {
    this._maxDepth = maxDepth;
  }

  subscribe(eventType: string, handler: EventHandler): void {
    if (!this._handlers.has(eventType)) {
      this._handlers.set(eventType, []);
    }
    this._handlers.get(eventType)!.push(handler);
  }

  async publish(event: ParsedEvent): Promise<void> {
    if (this._depth >= this._maxDepth) {
      console.warn(`Max publish depth ${this._maxDepth} reached, dropping event: ${event.type}`);
      return;
    }

    const handlers: EventHandler[] = [
      ...(this._handlers.get(event.type) ?? []),
      ...(this._handlers.get("*") ?? []),
    ];

    if (handlers.length === 0) {
      return;
    }

    this._depth++;
    try {
      const results = await Promise.allSettled(
        handlers.map((h) => h(event, this))
      );
      for (const r of results) {
        if (r.status === "rejected") {
          console.warn(`Handler error for ${event.type}: ${r.reason}`);
        }
      }
    } finally {
      this._depth--;
    }
  }
}

File: src/python/message_bus.py
from __future__ import annotations

import asyncio
import logging
from collections import defaultdict
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Callable

logger = logging.getLogger(__name__)


@dataclass
class ParsedEvent:
    type: str
    data: dict[str, Any]
    stream: AsyncIterator[str] | None = None
    raw: str | None = None


class MessageBus:
    def __init__(self, max_depth: int = 10) -> None:
        self._handlers: dict[str, list[Callable]] = defaultdict(list)
        self._max_depth = max_depth
        self._depth = 0

    def subscribe(self, event_type: str, handler: Callable) -> None:
        self._handlers[event_type].append(handler)

    async def publish(self, event: ParsedEvent) -> None:
        if self._depth >= self._max_depth:
            logger.warning(
                "Max publish depth %d reached, dropping event: %s",
                self._max_depth,
                event.type,
            )
            return

        handlers = list(self._handlers.get(event.type, []))
        handlers.extend(self._handlers.get("*", []))

        if not handlers:
            return

        self._depth += 1
        try:
            results = await asyncio.gather(
                *[h(event, self) for h in handlers],
                return_exceptions=True,
            )
            for r in results:
                if isinstance(r, Exception):
                    logger.warning("Handler error for %s: %s", event.type, r)
        finally:
            self._depth -= 1
